### PR TITLE
Fisher Information for Instruments

### PR DESCRIPTION
**Pull request by coreyostrove**
TLDR: This PR updates the implementation of the fisher information for experiment designs to support models/circuits with quantum instruments. This was requested ages ago in #464.

As part of updating the fisher information related code to support instruments this PR also changes the way that we do regularization. Previously this regularization was managed by adding a small amount of depolarizing error to the SPAM operations. Following a recommendation from @robinbk this has been changed to instead perform regularization by setting a minimum value for the probability p_min such that any probability below this threshold is clipped to this minimum value. As such, while in my testing the results looked nearly identical to previously calculated fisher information matrices for known experiment designs, the new code won't exactly reproduce the results of the older implementation.

### CODE REVIEW

**Review by sserita — APPROVED** [review]
No changes needed, thanks for these updates. Also nice catch on the `cumulative` flag not being used properly...
